Basic post functionality

thecatfour opened the pull request:
### Checklist

- [x] Feature
- [ ] Issue Fix
- [ ] Test
- [x] Domain
     - [x] Frontend
     - [ ] Backend
     - [ ] Both
- [ ] Documentation
     - [ ] Comments
     - [ ] README
     - [ ] Other Documents
- [ ] Other

### Description

This PR is a basic implementation of the frontend for posts. The following features regarding posts are addressed in this PR:

- Likes/Dislikes API call
- Displaying a list of posts
- Displaying the details of a specific post
- Deleting a post
- Creating a post

### Related Issues

This PR addresses issue #25.

### New Additions

- **Posts directory in app/:** Needed to display pages for the posts
- **Buttons:**
     - **Create Post:**  Adds a component that can redirect users to the create post page
     - **Delete Post:** Adds a component to delete a post
     - **Like/Dislike Post:** Adds a component to vote on a post
- **Delete Modal:** Adds a confirmation to the delete post button
- **Components:**
     - **Post Input:** Allows dynamic input for the post creation form
     - **Post List:** Allows the display of a list of posts
     - **Post Summary:** Allows basic information about a post to be displayed
- **Utils**:
     - **Input Helper:** Helper functions to check for errors
     - **Validations:** Constants for input validation for post creation
     - **Definitions:** Stores the Post type
     - **URL Segments:** Single file to store URLs

### Modifications

The app/, components/, and utils/ directories now have a posts/ directory to store the new files.

next.config.mjs now has the output key commented out. This was done to make dynamic routing for posts easier.

### New Dependencies

Package.json:
- **clsx:** Used to dynamically change the class of the like/dislike buttons
- **react-hook-form:** Used to create error messages that load on the create post form

Many of the new additions also use TypeScript.

### Visuals

The following is a use case diagram of how a user will interact with the included components:

![Posts drawio](https://github.com/user-attachments/assets/b45a8ac9-4728-4a01-8d6f-839131725f54)

The following is a gif of how a user might interact with the frontend:

![posts](https://github.com/user-attachments/assets/90b87801-0a8b-491e-9581-5a4b34cb6d1d)

### Architecture of Changes

**Like/Dislike Buttons**
These buttons have been implemented using a single component by passing in functions that activate when the button is clicked. This allows the same component to be used multiple times. It should also be able to be reused for comments in the future.

**Delete Post Button**
The delete button uses a modal to confirm whether or not the user wants to delete a post. If the user is on the post that they want to delete, they are redirected to the posts page. Currently, the delete button is present on every post, so in the future, it will need to be changed to only render on posts that a user owns or posts in a hub that a user moderates.

**Create Post Button**
The create post button simply redirects the user to a post creation page.

**Create Post Page**
This page uses the CreatePostInput component to display error messages on the form. After the errors are resolved as well as any afterwards validation (that has not been implemented yet), the page posts the form.

**Detailed Post Page**
This page displays the PostSummary component followed by other information about the post and the comments (which are not implemented yet). The page requires an input of the post id in order to render the page.

**Posts Page**
This page displays the PostList component along with a placeholder space for hubs. This may be changed in the future depending on how we decide to integrate posts and hubs in the frontend.

**Create Post Inputs Component**
This component is meant to allow an easy way to create form inputs while displaying errors on the form itself. It requires specific inputs, and examples can be found in utils/posts/create-post-validations.ts. 

**Post List Component**
This component is meant to provide a standardized way of displaying multiple posts. A developer just has to pass in a list of posts, and the component will display them in rows. However, the component does not provide post filtering. The page that wants to display the list will need to filter the posts.

**Post Summary Component**
This component is meant to provide a standardized way of displaying basic information about a single post. It includes the like/dislike buttons as well as the delete button. It also has a thumbnail placeholder in case we want to implement an image thumbnail in the future. Clicking on the thumbnail or the title will redirect the user to the post's detailed page. The component takes a post as input and then displays some information about the post.

### Use of Changes

The PostList component can be used to display multiple posts for pages such as profiles or hubs. 

The PostSummary component can be used to display a single post.

The likes/dislikes buttons can potentially be reused for comments.

### Testing

**Issue 1**
When loading the posts page or a detailed post page, sometimes the page begins loading before the api call is finished. This results in a runtime error where the PostSummary or PostList component try to access a field in null. This has been temporarily solved by adding some if statements to validate the response data.

**Issue 2**
The likes and dislikes appear correct on the client's side, but they do not operate correctly on the backend. Currently, clicking like twice appears like the second click undoes the initial like. However, on the backend, the database adds two likes to the  post. This will need to be fixed on the backend. Additionally, if a user likes a post, refreshing the page will make it appear as if they did not like the post. One potential solution is creating a likes and dislikes database to store user likes and dislikes.

thecatfour committed:
Basic Posts Implementation

Added basic frontend functionality for posts.

thecatfour committed:
Merge branch 'master' into master

thecatfour committed:
Merge pull request #1 from UNLV-CS472-672/master

Updating Fork

thecatfour committed:
Minor packages fix

thecatfour committed:
Added message for empty post list.

thecatfour committed:
Revert "Added message for empty post list."

This reverts commit 9a9e2d30a2dc92d7cdb58a29863f14a724e96abb.

k

thecatfour committed:
Added a check for

thecatfour committed:
Changed how the dynamic parameter is received.

thecatfour committed:
Changed create post to redirect to new post.

thecatfour committed:
Changed ProtectedRoutes.

thecatfour committed:
Add some input validation

thecatfour committed:
Modified likes and dislikes

Only like or dislike can be selected on the frontend.

thecatfour committed:
Update .gitignore

thecatfour commented:
I modified the like/dislike component. Now only like or dislike can be chosen on the frontend. However, the backend still increments likes and dislikes each type the button is clicked.

![posts like and dislike](https://github.com/user-attachments/assets/c7309d8b-06c5-4827-b223-afe46d52f85e)

misherwa commented:
Hey Leslie, 

For issue 2 that you mentioned, I can create a new Issue for the like and dislike increment for Posts. I plan on fixing the like and dislike issue after submitting my backend Comments PR.

I was wondering how I would get to the same pages as your two gifs. I'm having issues testing the frontend. My URL address looks like this: https://possessed-spooky-haunting-697vg9xj5vg3wqr-3000.app.github.dev/

And this is what the page looks like:
![q2](https://github.com/user-attachments/assets/3f4fdfb3-e4a9-42d7-97b7-7b4dd34872ea)

I know you mentioned in your first issue that, "When loading the posts page or a detailed post page, sometimes the page begins loading before the api call is finished". So when I do click on the "[Go to posts]", it leads me to an error page. Is that what you were talking about in your first issue?

For some much smaller changes that you could include, I recommend adding comments a simple comment for both your page.js files in the "posts" folded and the one in the "create" folder. I feel like adding at least one comment to every file improves code clarity or is helpful. In this case helps to differentiate the two files even more.
![q](https://github.com/user-attachments/assets/9b52d340-21db-47c3-9696-61ee5918c109)

Correct me if I'm wrong, but it looks like the page.js in the "posts" folder is for viewing the posts, so you could add a single comment like, "Displays a list of posts for the user.". You can go into future detail if you want,

And for the page.js in "create" is for creating the posts, so you can put a comment like, "Allows users to submit new posts." and add more details if you want.

misherwa commented on frontend/src/components/posts/pages/create-post.tsx:
```diff
@@ -0,0 +1,62 @@
+            router.push(URL_SEGMENTS.FRONTEND + URL_SEGMENTS.POSTS_HOME + response.data.id);
+
+        } catch (error) {
+            alert(error);
```
I feel like you should include some sort of error message to this if creating a new post fails. It could be like: 
setErrorMessage(error.message ||"An Error has occurred, so please try again.");

misherwa commented:
Hey Leslie, 

This recommendation is for something that would be nice to have. You could probably implement this in a later PR, but it's by adding a variable called "isLoading". Basically, it's a boolean variable that can be used when fetching data. This might help with Issue 1, but I'll try to test it after I get the page running.

It is similar to how you get that loading screen when something is loading, like on a website or video. So it seems like it's possible to include this in "create-post.tsx", "detailed-post.tsx", and posts.tsx. All these files can be found in frontend/src/components/posts/pages

But I was hoping that you could implement it whenever fetching data from the API or needing to wait for things to update.

Here are some links to get a better idea of it, I recommend checking out the first one:
https://www.mariokandut.com/how-to-handle-errors-and-data-loading-state-with-react-hooks/

and:
https://www.shecodes.io/athena/11556-react-how-to-show-a-loading-message-when-fetching-data

thecatfour commented:
@misherwa  

Regarding the issue you are having when clicking the posts/ page, I believe this is the ProtectedRoute component redirecting you to a non-existent page. When you click on the link, the URL probably displays something like .../posts/users/login, which is not a valid page. In order to access posts, you need to log in first at http://localhost:3000/users/login/. This gives you the authentication that is required to view posts.

Also, thanks for the "isLoading" suggestion. The pages that I currently have set up may not be used in the actual website, so it may be better to implement that idea in a future PR. I mostly made the current posts pages to help show how the frontend post elements work.

thecatfour committed:
Create DetailedPost component

thecatfour committed:
Added simple comments to pages

thecatfour commented:
I created a new component to display a detailed post rather than relying on a specific page to do so. The three main reusable components are now PostList, PostSummary, and PostDetailedSummary. PostList takes an array of posts while the other two take a post and display information.

I also added simple comments to the pages in the app/ directory.

barrioskevin commented:
Hello, I really like how you setup the posts view and handled the lists. I think the changes you added are fine and help the project by adding in the pieces needed for viewing posts. I believe what you have added will be needed once we start putting the main app's pages together. 

Just for how you designed the post detailed view I'm not sure the bubbles for the individual details work as good. Maybe the details should come out smaller so that the main focus is on the posts message and comments. But that's just a nitpick.
![image](https://github.com/user-attachments/assets/f9b8a219-6555-4411-aeef-51c70a9bce92)

It took me a while to understand, but the visual you provided is a good way to show the relationship between the different components. If you focused on showing that I believe the visual could have been more clear but thanks for providing one.

Also the changes you made to page.js we can probably leave out for now. The button you added, when you click it and you are not signed in, will attempt to take you to the login page but it still adds in your posts to the url so we end up going to `posts/users/login` and get a 404. This is most likely a bug in `ProtectedRoute.js` where we don't properly define the url to go to when a user is not logged in.
![image](https://github.com/user-attachments/assets/992c09a2-150d-4a87-84f8-b02a0183be71)

So I think it's best we don't edit page.js until we start working on that posts timeline.

A lot of new tsx files but I was able to run the project no problem so things should be fine when the workflow eventually gets added. 

When the button you added on page.js is clicked when a user IS logged in, I also think it glitches on the first render well at least for me.
![firstrender](https://github.com/user-attachments/assets/a82a3397-6ac6-4c46-a967-810fda7d56ed)

Things go to normal after a refresh tho. 

In my opinion you should just remove the changes to `page.js` and if you want patch `ProtectedRoute.js` or just remove your changes its a small little change so you can just remove the change if you don't want to mess with it we can open an issue for that redirect thing.

Other than that I don't really see much issues with your changes and think you did a really nice job this should be a good step for the progression of the project.

thecatfour committed:
PR changes

thecatfour commented:
Changes:

- The redirect on ProtectedRoute now redirects to the login page regardless of the URL
- Changed app/page.js to remove the link to posts
- Changed the text on the buttons, so they should be visible again

I did not change the strange first render bug. We can probably fix this when we actually start building the main pages.

barrioskevin commented:
Thanks for fixing the ProtectedRoute component. Removing the changes to page.js was also a good call, the main focus can now be using what we have to make page.js our actual page.
